Use bigint for large numbers

Gouvernathor opened the pull request:
Built on top of #40.
It is disabled by default. If the parse function is passed NUMBER_OR_ERROR directly (or if it is not passed an integer parsing option), the return type of the function will not include bigint, so it is backwards-compatible.
Fixes #31.

The new features are not covered by tests, but the existing tests (including internal unit tests) have been updated following general internal refactors.

Gouvernathor committed:
Use more detailed TOML types

the types themselves are inspired from type-fest's json types

Gouvernathor committed:
semicols

Gouvernathor committed:
Implement conditional bigint parsing

the default remains the same, and the return type errors on calls

Gouvernathor committed:
Add alternate signatures

Gouvernathor committed:
Add bigint to the possible primitive types

Gouvernathor committed:
use undefined instead of the void operator

Gouvernathor committed:
Pass the integerParsing option through

Gouvernathor committed:
Simplify the parseValue signature

Gouvernathor committed:
Fix signature of the parse function

Gouvernathor committed:
Reimplement types to allow changing the leaves

Gouvernathor committed:
Add more precise typing to the parse function through override signatures

Gouvernathor committed:
Tweak

Gouvernathor committed:
Tweak signatures (and location of the default) to help test adaptation

Gouvernathor committed:
Visually separate the default parse and the number_or_error parse

if the default integer parsing is changed in a future version, only the first signature's return type will need to change, rather than a more complex edit

Gouvernathor committed:
Fix tests

Gouvernathor committed:
Export the TOMLTable too

It is the return type of the parse function, so it should be exposed
I'm against exposing them all

cyyynthia commented on src/primitive.ts:
```diff
@@ -126,51 +126,66 @@ export function parseString (str: string, ptr = 0, endPtr = str.length): string
 }
 
-export function parseValue (value: string, toml: string, ptr: number): boolean | number | TomlDate {
+export enum IntegerParsing { NUMBER_OR_ERROR, NUMBER_OR_BIGINT, BIGINT_ONLY }
```
Exposed enums is kind of a footgun, as the TS output is relatively huge (and quite gross). I generally just use a string union, not the prettiest but it's the most straightforward.

Alternatively, it can be a `boolean | 'as_needed'`, which allows to treat false (`!integerParsingMode`) as "error on precision loss", and `integerParsingMode === true` to be "always bigint", where both do minify quite well.

Gouvernathor replied:
I like how string unions are supported by Typescript. It's also clearer to use from javascript, and I don't like boolean-or-else types : for me a boolean is meant to be binary.

cyyynthia replied:
Sometimes they're expressive enough... I agree as a type theorist (using an enum is the proper solution, string union is a poor man's enum so it's equally acceptable), but as a TS dev optimizing for code-size... The fact undefined is falsey allows for neat optimization tricks 🤷🏻‍♀️

cyyynthia commented on src/primitive.ts:
```diff
@@ -126,51 +126,66 @@ export function parseString (str: string, ptr = 0, endPtr = str.length): string
-export function parseValue (value: string, toml: string, ptr: number): boolean | number | TomlDate {
+export enum IntegerParsing { NUMBER_OR_ERROR, NUMBER_OR_BIGINT, BIGINT_ONLY }
+
+export function parseValue(value: string, toml: string, ptr: number, integerParsing: IntegerParsing = IntegerParsing.NUMBER_OR_ERROR): boolean | number | bigint | TomlDate {
```
Default value is not needed, this is an internal function.

Gouvernathor replied:
The issue is that it makes the unit tests more verbose, since in that case all functions except for parse will need the parameter to be passed. Is that what you want ?

cyyynthia replied:
Yeah, at least the tests will be explicit about what's the precise thing being tested. Better to have the extra weight in the tests rather than in production bundles :)

Gouvernathor replied:
Done

cyyynthia commented on src/primitive.ts:
```diff
@@ -126,51 +126,66 @@ export function parseString (str: string, ptr = 0, endPtr = str.length): string
+	if (value === 'nan' || value === '+nan' || value === '-nan') return NaN;
 
-	if (value === '-0') return 0 // Avoid FP representation of -0
+	if (value === '-0') return 0; // Avoid FP representation of -0
```
This path must not be taken when we always use bigints.

Gouvernathor replied:
Done, but how about `-0.0` ? That's not handled, is it ?

Gouvernathor replied:
In fact there's even a test that "-0.0" returns the negative zero. Shouldn't that be fixed so that "-0" always returns the same thing as "-0.0" ?

cyyynthia replied:
This special case was done so that `-0`, `0`, and `+0` are strictly equal to `0`. This is generally intended, and explicitly specified by the specification: https://toml.io/en/v1.0.0#integer

While the floating point `-0.0` is equal to `0.0` per IEEE 754, its behavior differs in some edge-cases. They're equal (`-0 === 0` is true), but not strictly equal (`Object.is(-0, 0)` is false, `1/0` is Infinity but `1/-0` is -Infinity). So, since `-0` would get parsed as a float, there needs to be a special case for it such that it does return exactly `0`.

Gouvernathor replied:
I'm adding a test for that.

cyyynthia commented on src/primitive.ts:
```diff
@@ -126,51 +126,66 @@ export function parseString (str: string, ptr = 0, endPtr = str.length): string
 	// Numbers
-	let isInt
-	if ((isInt = INT_REGEX.test(value)) || FLOAT_REGEX.test(value)) {
+	const isInt = INT_REGEX.test(value);
+	const isFloat = FLOAT_REGEX.test(value);
```
I use `let` because it's shorter than `const`, I think most if not all variables are `let` and not `const`.

Also, storing both is wasteful: given that we enter the branch only of it is an integer or a float, if `isInt` is false then you know it must be a floating point value. This is why it was written this way before, and I don't think modifying it is necessary at all.

Gouvernathor replied:
It's not wasteful because I use it in another if. (I wouldn't waste a variable otherwise, I agree.) But it may disappear in another refacto.

Gouvernathor replied:
Done

cyyynthia replied:
My point is that if you know `isInt`, then you necessarily know `isFloat` when you're inside the numeric processing logic. `isFloat === !isInt`, so there's no need to store it alongside `isInt`

cyyynthia commented on src/primitive.ts:
```diff
@@ -126,51 +126,66 @@ export function parseString (str: string, ptr = 0, endPtr = str.length): string
+			});
 		}
 
-		let numeric = +(value.replace(/_/g, ''))
```
The only special handling comes from integer values. So I think quite a lot can be simplified.

```ts
value = value.replace(...)
let numeric = +value
if (isNaN(value)) ...

if (isInt) {
  // isInt is no longer used, so we can reuse it to keep whether it's a safe int or not.
  if (!(isInt = Number.isSafeInteger(numeric)) && !useBigIntegers) ... // Precision loss.
  if (!isInt || useBigIntegers == 1) numeric = BigInt(value)
}
```

Gouvernathor replied:
We don't even need to reuse the variable, actually (I do frown upon that)

cyyynthia replied:
Yeah I'd frown upon that in most other situations too. Again, here I'm in the shoes of someone optimizing for code size, and sometimes sacrifices needs to be made. It's not *too* bad either, as the type remains the same at least. 🤷🏻‍♀️

I'm sure you've seen worse in the code already, I used some very cursed tricks here and there 😅

Gouvernathor replied:
TODO : I think that version doesn't support numbers over e310, since once converted to number they are nan but yet they may be integers supported by bigint.
There needs to be a test about that too.

cyyynthia replied:
`1e1` forms are floating point numbers, never integers. https://toml.io/en/v1.0.0#float

Gouvernathor replied:
I meant higher than that value, they can be written as a string of 300+ digits

cyyynthia replied:
Are you sure they get converted to NaN? I get Infinity after a quick test, which is fine (it's not NaN, and it's not a safe integer either) 🤔 

```
Node REPL
> +('9'.repeat(310))
Infinity
```

Gouvernathor replied:
I added tests for that.

cyyynthia commented on src/parse.ts:
```diff
@@ -113,10 +114,15 @@ function peekTable (key: string[], table: Record<string, TomlPrimitive>, meta: M
-	let maxDepth = opts?.maxDepth ?? 1000
-	let res = {}
-	let meta = {}
+export function parse(toml: string, options?: { maxDepth?: number }): TOMLTable<Exclude<TOMLPrimitive, bigint>>;
+export function parse(toml: string, options?: { maxDepth?: number, integerParsing: IntegerParsing.NUMBER_OR_ERROR }): TOMLTable<Exclude<TOMLPrimitive, bigint>>;
+export function parse(toml: string, options?: { maxDepth?: number, integerParsing: IntegerParsing }): TOMLTable;
```
The generic options type can be exported as `ParseOptions`, and the overloads can be written with `ParseOptions & { integerParsing: ... }`

Gouvernathor replied:
The first overload can't be expressed like this - or I don't know how

cyyynthia replied:
The order of the overloads matter. [TypeScript playground](https://www.typescriptlang.org/play/?#code/C4TwDgpgBA8mwEsD2A7AzlAvFA3gKCkKgEMB+ALijWACcEUBzAogIwqhaSQBsJiU8AXzx4AZgFcUAY0SooAEyQBlJAFsIwABb0GUABTFKcWeigAyXB0q1x0QQEpKABjGSZyFAuVqN2xvsNYeA8MCxwrKFFibjQ7RygARldpEy8VdS0dAKNg1DR4lHFVFggaZPc5RXTfLIMck3zKQuLS3GZCGg1xGk8ATiERKTzgEiw0n0zGPRwHPAB6OaIoAD1SPCH0EZYxqom-BmmImzj5xaJV9eGoKR3vDP3DlkoomJOFpYugA)

Gouvernathor replied:
Is this what you have in mind ?
```ts
export function parse (toml: string, options?: ParseOptions & { integerParsing: "number_or_error" }): TomlTable<Exclude<TomlPrimitive, bigint>>;
export function parse (toml: string, options?: ParseOptions & { integerParsing: IntegerParsing }): TomlTable;
export function parse (toml: string, options?: ParseOptions): TomlTable<Exclude<TomlPrimitive, bigint>>;
```
I don't find it very readable personally.

cyyynthia replied:
Hm, it can be done in 2 overloads only, i.e.

```ts
export function parse (toml: string, options?: ParseOptions & { integerParsing: Exclude<IntegerParsing, 'number_or_error'> }): TomlTable
export function parse (toml: string, options?: ParseOptions): TomlTable<Exclude<TomlPrimitive, bigint>>
```

It's not ideal, but as `ParseOptions` will eventually grow it'll become even more of a pain. Intersected type could be hoisted, i.e.

```ts
type TomlTable = { [key: string]: ... }
type TomlTableWithBigint = { [key: string]: bigint | ... }

// ---

type ParseOptionsWithBigInt = ParseOptions & { integerParsing: Exclude<IntegerParsing, 'number_or_error'>

export function parse (toml: string, options?: ParseOptionsWithBigInt): TomlTableWithBigInt
export function parse (toml: string, options?: ParseOptions): TomlTable
export function parse (toml: string, options?: ParseOptions): TomlTableWithBigInt {
  ...
}
```

Gouvernathor replied:
I did it the first way. I don't generally like single-use definitions.

cyyynthia commented on src/struct.ts:
```diff
@@ -116,8 +116,11 @@ export function parseKey (str: string, ptr: number, end = '='): [ string[], numb
-	let res: Record<string, TomlPrimitive> = {}
+export function parseInlineTable(
+	str: string, ptr: number,
+	{ depth = -1, integerParsing }: { depth?: number, integerParsing?: IntegerParsing|undefined } = {}
```
!! objects as arguments are nice in some places, but very bad here as it prevents minifiers from making everything a single-letter variable. Getting rid of the -1 default would be ideal :/

Gouvernathor replied:
Except for exceptionally stable interfaces (such as string.substring for instance), I think all optional parameters should be in keyword option parameters.
But even short of that, I think positional-only parameters are bad as soon as two independent parameters are optional - as was visible from the use of `void 0` somewhere (maybe in tests) to skip a parameter.
We can make depth required and integerParsing the only optional, but the same problem would bite us back as soon as the next option is implemented...

cyyynthia replied:
Those functions are internal and won't ever be exposed to public-facing APIs, and the code is small enough that dealing with interfaces changing is a non-issue. The code size impact does matter a lot to me. It must be *smol*.

All parameters can be marked as required, again since it's internal and the fact they're optional is only something the user sees in the public APIs.

Gouvernathor replied:
extractValue's `end` parameter is now required with undefined being allowed, which is slightly regrettable but better than putting it away from ptr.

cyyynthia reviewed (changes requested):
Thanks for the PR! Stylistic nits mentioned in #40 apply - please try to keep the original code style intact (I know, no formatter, annoying, sorry!) 🙂

Gouvernathor committed:
Recase (and reorder)

Gouvernathor committed:
Spaaaaaaaaaace

Gouvernathor committed:
Mark deprecation

Gouvernathor committed:
Turn the enum into union of strings (also expose the type)

Gouvernathor committed:
Refacto

Gouvernathor committed:
Remove readonly array, and refactor out TomlArray

Gouvernathor committed:
Merge branch 'mistress' into bigint

Gouvernathor committed:
Explicit tests

Gouvernathor committed:
Remove internal named parameters

Have to do it all in one go, since they call one another recursively
The end parameter now needs to be explicitly passed

Gouvernathor committed:
Fix tests

Gouvernathor committed:
Handle the 0/-0 case

Gouvernathor committed:
Refacto

Gouvernathor committed:
Add tests for signed integer zero

Gouvernathor committed:
Export the options type by itself

Gouvernathor committed:
Drop the object, keep only the type

Gouvernathor committed:
Add tests

Gouvernathor committed:
Remove undefined in TomlTable

Gouvernathor committed:
Merge branch 'mistress' into bigint

Gouvernathor commented:
This, in the readme 

> floats will be serialized as integers if they don't have a decimal part. `stringify(parse('a = 1.0')) === 'a = 1'`

could be disabled in an option on the stringify function, to serialize all numbers as floats, and it would mirror the integers_as_bigint option on parse. That way, you could have type-stable parse-serialization cycles.

cyyynthia commented on test/value.test.ts:
```diff
@@ -32,149 +32,172 @@ import { TomlError } from '../src/error.js'
+
+it('supports numbers larger than the max safe integer number when bigints are enabled', () => {
+	expect(parseValue('9007199254740992', '', 0, "number_or_bigint")).toBe(BigInt('9007199254740992'))
+	expect(parseValue('9007199254740992', '', 0, "integers_as_bigints")).toBe(BigInt('9007199254740992'))
```
Style: expected bigint could be written using `0n` syntax instead

cyyynthia reviewed (commented):
I have stylistic nits, but nothing too important. I only annotated one but I could also mention the presence of semicolons etc. Tests look good, I'll probably change a few things in the implementation before release but the core logic seems good.

You don't necessarily need to fix style issues, I can handle that myself before merging. I'll squash and merge PRs, so they'll need to be rebased once I start merging things.

Gouvernathor committed:
Use index type

Gouvernathor committed:
use the deprecated tag

Gouvernathor committed:
Merge branch 'mistress' into bigint

Gouvernathor committed:
Use better syntax

Gouvernathor committed:
Fix incorrect testing

Gouvernathor committed:
no-call bigint syntax

Gouvernathor committed:
Add test

Gouvernathor committed:
Merge branch 'mistress' into bigint

Gouvernathor committed:
Moar test

Gouvernathor committed:
That's const bru

cyyynthia committed:
chore: formatting nits, refactor bigint opt

cyyynthia committed:
chore: consistent casing of BigInt
